### ISSUE TITLE
Picking of stock allows multiple picks from same bin

### ISSUE THREAD

**Issue by MischaSLabbert** [openboxes/openboxes#229]
This is part of the picking of an outbound shipment. When you have multiple lines (same item and lot scattered over multiple bins) the system allows you to pick from the same bin over and over, regardless of the fact that it was picked previously, and that there is theoretically not enough stock to fulfill the shipment. It also allows you to send the shipment, without throwing an error. 

To help here is an example.
First a view of our inventory levels:

![image](https://user-images.githubusercontent.com/34237733/33551572-720da9f6-d8fa-11e7-8e6d-9ce794533c71.png)

We create the outbound, and edit the packing list. The system does throw the error correctly if you have selected a batch and you do not have the correct qty’s. in the example I have selected to ship all the stock.
![image](https://user-images.githubusercontent.com/34237733/33551593-899e2f0a-d8fa-11e7-9633-deddefd63762.png)

Moving to the “Pick shipment items”, the system allows me to pick all the stock from the stock with a qty of 93 (I have done this, to demonstrate what’s happening.)

![image](https://user-images.githubusercontent.com/34237733/33551607-988fb90c-d8fa-11e7-969d-256ba161efa0.png)

As you can see I have selected the stock all from location Bin2, without getting a error. Even “Validate Picklist” does not throw an error.
As you can see, I am able to ship the shipment

![image](https://user-images.githubusercontent.com/34237733/33551632-a6b6a202-d8fa-11e7-9819-d26b7f4084be.png)

And my inventory is as follow:
![image](https://user-images.githubusercontent.com/34237733/33551640-afc6b526-d8fa-11e7-9a4e-48acddb56b0c.png)

**Comment by jmiranda** [comment]
Thanks guys. I'm going to have some time to look into this later this week.

**Comment by jmiranda** [comment]
Confirmed bug.

**Comment by jmiranda** [comment]
TLDR; The system has no notion of allocation -- that is why we're running into this problem.

So this is actually more of feature enhancement than a bug, but definitely still a bug since we don't have a notion of "allocation" yet.  In other words, we're only doing quantity validation for each shipment item, not all shipment items together. Therefore, the quantity for each of these shipment items validates to true since nothing has actually been allocated yet (i.e. 90 EA, 45 EA, and 93 EA are all less than or equal to 93 EA). Obviously, added together (228) the quantity should validate to false since it's more than 93.

This is basically what the code is doing -- it's basically taking each individual shipment item and comparing against quantity on hand.
```
boolean validatePicklist(Shipment shipment) {
    shipment.shipmentItems.each { shipmentItem ->
        def quantityOnHand = getQuantityOnHand(shipmentItem)
        if (shipmentItem.quantity > quantityOnHand) {
            throw ValidationException("Quantity on hand must be greater than item quantity")
        }
    }
    return true
}
```
Instead we need to take allocated stock into account as well
```
boolean validatePicklist(Shipment shipment) {
    shipment.shipmentItems.each { shipmentItem ->
        def quantityAvailable = getQuantityOnHand(shipmentItem) - getQuantityAllocated(shipmentItem)
        if (shipmentItem.quantity > quantityAvailable) {
            throw ValidationException("Quantity on hand must be greater than item quantity")
        }
    }
    return true
}
```
So the question is, how should we build in the notion of allocation? Do we mark an item as "allocated" once it is "picked" for a shipment? This would be the most straightforward path. The other question is what happens when we have multiple outbound shipments that are all in the picking process. You might have picking/allocation conflicts between the shipments, which doesn't really change the allocation problem (this still straightforward) but it complicates the picking process a bit. Just something to keep in mind since it makes the picking process a little more complicated. How do we make it less painful? 

Also note that the outbound picking process is not quite where I want it to be. At the moment we're just re-using the shipment item to track picked quantities. In fact, we're using the original quantity field, so if you requested 100 EA to be shipped and then picked 80 EA the shipment.quantity = 80. There's no quantity_picked or a separate picklist component yet.

This has been bothering me since we launched this feature but there's not much I could do at the time. At some point in the near future, I'd like to add in the "picklist" component to the shipment picking process so we can avoid this. Then I'd be able to handle all "allocation" calculations by looking at the picklist and not checking the quantity and status of each shipment item.

Anyway, let me know your thoughts. Getting a quick and dirty fix in should be easy, but just wanted to add in a few questions so that we consider the big picture.

**Comment by jmiranda** [comment]
![pick shipment items 3](https://user-images.githubusercontent.com/1162935/33840248-fc80f656-de59-11e7-9ebf-1b52a6b2b506.png)

**Comment by jmiranda** [comment]
Should be fixed in 0.7.26
